ts/keys.go

```diff
@@ -0,0 +1,114 @@
+// perfect resolution sample would keep the original timestamp and value for
+// every data point; lower resolution samples aggregate data points which fall
+// within the same sample period, keeping only the sum and count of all data
+// points. Multiple resolutions may be kept for a single metric; lower
+// resolutions are used to quickly query larger timespans, while higher
+// resolutions are used to provide more detail for smaller timespans. The
+// resolution is denoted in the key by a three character string.
+//
+// The timestamp is the first timestamp for which this key might contain data.
+// Multiple data samples are stored at the same key; a typical configuration
+// might have a ten second sample period, with all samples for the same hour
+// stored in the same key.
+//
+// The source is a possibly empty string which identifies the source for a
+// metric.  For example, a metric may describe an overall statistic for a
+// cluster of machines, but the data from each individual machine is stored
+// separately using a different source name. The source key is appended to the
+// end of the key; as such, keys for different sources with the same
+// metric name and timestamp will sort adjacently in the global key value store.
+// In this way, a metric can be quickly queried across all sources.
+var (
+	// KeyDataPrefix is the key prefix for time series data keys.
+	KeyDataPrefix = proto.MakeKey(engine.KeySystemPrefix, proto.Key("tsd"))
+	// KeyResolutionLength is the required length of resolution key components.
+	KeyResolutionLength = 3
+	// KeyResolutionTenSeconds is the resolution key component for keys which
+	// store data with a ten second sample duration.
+	KeyResolutionTenSeconds = proto.Key("10s")
+)
+
+// MakeDataKey creates a time series data key for the given metric name,
+// source, resolution and timestamp.
+func MakeDataKey(name string, source string, r Resolution,
+	timestamp int64) proto.Key {
+	return proto.MakeKey(KeyDataPrefix, encoding.EncodeBytes(nil, []byte(name)),
+		r.KeyComponent(), encoding.EncodeVarint(nil, timestamp),
+		encoding.EncodeBytes(nil, []byte(source)))
+}
+
+// DecodeDataKey decodes a time series key into its components.
+func DecodeDataKey(key proto.Key) (string, string, Resolution, int64) {
+	var (
+		name       []byte
+		source     []byte
+		resolution Resolution
+		timestamp  int64
+		remainder  = key
+	)
+
+	// Detect and remove prefix.
+	if !bytes.HasPrefix(remainder, KeyDataPrefix) {
+		panic(fmt.Sprintf("malformed time series data key %v: improper prefix", key))
+	}
+	remainder = remainder[len(KeyDataPrefix):]
+
+	// Decode metric name.
+	remainder, name = encoding.DecodeBytes(remainder)
+
+	// Decode resolution.
+	if len(remainder) < KeyResolutionLength {
+		panic(fmt.Sprintf("malformed time series data key %v: no resolution", key))
+	}
+	resolution = ResolutionForKeyComponent(remainder[:KeyResolutionLength])
+	remainder = remainder[KeyResolutionLength:]
```

Comment:
Curious why you are encoding the resolution as a string. I would have thought to encode it as a varint.